### PR TITLE
Add Codable Support for Swift 4.x

### PR DESCRIPTION
**Pull request by csantanapr**
## Codable Support with Swift 4.x

Some examples of using Codable In and Out

### Codable style function signature
Create file `helloCodableAsync.swift`
```swift
// Domain model/entity
struct Employee: Codable {
  let id: Int?
  let name: String?
}
// codable main function
func main(input: Employee, respondWith: (Employee?, Error?) -> Void) -> Void {
    // For simplicity, just passing same Employee instance forward
    respondWith(input, nil)
}
```
```
wsk action update helloCodableAsync helloCodableAsync.swift swift:4.1
```
ok: updated action helloCodableAsync
```
wsk action invoke helloCodableAsync -r -p id 42 -p name Carlos
```
```json
{
    "id": 42,
    "name": "Carlos"
}
```

### Codable Error Handling
Create file `helloCodableAsync.swift`
```swift
struct Employee: Codable {
    let id: Int?
    let name: String?
}
enum VendingMachineError: Error {
    case invalidSelection
    case insufficientFunds(coinsNeeded: Int)
    case outOfStock
}
func main(input: Employee, respondWith: (Employee?, Error?) -> Void) -> Void {
    // Return real error
    do{
        throw VendingMachineError.insufficientFunds(coinsNeeded: 5)
    } catch {
        respondWith(nil, error)
    } 
}
```
```
wsk action update helloCodableError helloCodableError.swift swift:4.1
```
ok: updated action helloCodableError
```
wsk action invoke helloCodableError -b -p id 42 -p name Carlos
```
```json
{
"name": "helloCodableError",
"response": {
  "result": {
    "error": "insufficientFunds(5)"
  },
"status": "application error",
"success": false
}
```

### CODE REVIEW

**Comment by csantanapr** [comment]
Depends on #22 
Closes #2

**Comment by csantanapr** [comment]
TODOs

- [ ] Handle when input params is empty
- [ ] Add tests cases for Codable

**Comment by jthomas** [comment]
I've been testing out Codable support from the sample above in my library for wrapping Swift actions to produce binaries.
https://github.com/jthomas/OpenWhiskAction/blob/swift4/Sources/epilogue.swift

It's all working great but I made a few changes to improve the developer experience that we'd probably want to include. Here are the things I found....

- Support `@escaping` closures to allow users to use with async calls.
```
func _run_main<In: Codable, Out: Codable>(mainFunction: (In, @escaping (Out?, Error?) -> Void) -> Void) {
}
```

This just needs the annotation adding and doesn't affect existing code or people who don't use that.

- Improved error handled to give more descriptive messages, e.g.
https://github.com/jthomas/OpenWhiskAction/blob/swift4/Sources/epilogue.swift#L84-L93

This helps the user understand when there has been problems using the Codable API versus something else.

- Sync method style APIs should probably support throws. This enables uses to use normal error handling to surface issues.

I'm away on vacation for a week but will attempt to push a PR when I return.

**Comment by csantanapr** [comment]
Thanks @jthomas for the review and starting the work to externalize the epilogue in to it's own library/package.


>Support @escaping closures to allow users to use with async calls.
```
func _run_main<In: Codable, Out: Codable>(mainFunction: (In, @escaping (Out?, Error?) -> Void) -> Void) {
}
```
This just needs the annotation adding and doesn't affect existing code or people who don't use that.

Guess what I have in my TODOs? 

### TODOS:
- [ ] Finish with the Codable test cases
- [ ] Use `@escaping` instead for the callback
- [ ] Refresh again with latest 4.1 snapshot
- [ ] With now 4.2 announced 4.1 is stable, so skip/remove 4.0 support only do 4.1 for now

>https://github.com/jthomas/OpenWhiskAction/blob/swift4/Sources/epilogue.swift#L84-L93
This helps the user understand when there has been problems using the Codable API versus something else.

I will add the one for more verbose errors 👍 
- [ ] Improved error handled to give more descriptive messages, e.g.

>Sync method style APIs should probably support throws. This enables uses to use normal error handling to surface issues.

I thought about this one, will look into this but probably a different PR since it's an improvement to dictionary sync that we have today.
https://github.com/apache/incubator-openwhisk-runtime-swift/issues/29

Super happy that someone is trying this stuff 🎉

**Comment by csantanapr** [comment]
Woot !

- [ ]  Finish with the Codable test cases

**Comment by csantanapr** [comment]
Woot !

- [ ] Refresh

 again with latest 4.1 snapshot

**Comment by csantanapr** [comment]
Woot !

- [ ]  Improved error handled to give more descriptive messages, e.g.

**Comment by csantanapr** [comment]
> With now 4.2 announced 4.1 is stable, so skip/remove 4.0 support only do 4.1 for now

Going to move this to a new issue https://github.com/apache/incubator-openwhisk-runtime-swift/issues/30

**Comment by csantanapr** [comment]
I think I'm ready for review and final merge of this PR

File: README.md [hunk 1]
```diff
@@ -39,6 +39,76 @@ func main(args: [String:Any]) -> [String:Any] {
...
+"response": {
+  "result": {
+    "error": "insufficientFunds(5)"
+  },
+"status": "application error",
+"success": false
+}
+```
+
 ### Packaging an action as a Swift executable using Swift 4
```
Comment (paulcastro) [inline README.md]: Line 209: `All your source code needs to be copied ... `
Line 231: `Take into account that starting ...`

File: core/swift40Action/epilogue.swift [hunk 2]
```diff
@@ -21,24 +21,100 @@ import Foundation
...
+        if JSONSerialization.isValidJSONObject(result) {
+            do {
+                let jsonData = try JSONSerialization.data(withJSONObject: result, options: [])
+                if let jsonStr = String(data: jsonData, encoding: String.Encoding.utf8) {
+                    print("\(jsonStr)")
+                } else {
+                    _whisk_print_error(message: "Error serializing data to JSON, data conversion returns nil string", error: nil)
+                }
+            } catch {
+                _whisk_print_error(message: "JFailed to encode Dictionary type to JSON string:", error: error)
```
Comment (paulcastro) [inline core/swift40Action/epilogue.swift]: What's `JFailed` ?  Typo?

Comment (csantanapr) [reply core/swift40Action/epilogue.swift]: yeah Typo

File: README.md (hunk 1 above)
Comment (csantanapr) [reply README.md]: updated the readme with typos fixes @paulcastro

**Pull request merged**